### PR TITLE
Epoch nodes config cacher

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
- Cacher for epoch nodes config, in order to keep more nodes configuration in a cacher if needed.
  
## Proposed changes
- add a separate cacher for epoch nodes config
- separate operation for getting nodes config

## Testing procedure
- Standard system test

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 1]
```diff
@@ -226,10 +229,46 @@ func checkArguments(args ArgNodesCoordinator) error {
...
 		return ErrInvalidNumberOfStoredEpochs
 	}
+	if check.IfNilReflect(args.NodesConfigCache) {
+		return ErrNilNodesConfigCacher
+	}
 
 	return nil
 }
 
+// getNodesConfig will try to get nodesConfig from map, if it doesn't success, it will try to get it from nodes config cache
```
Comment (AdoAdoAdo) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:239]: if it doesn't succeed, ..

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 2]
```diff
@@ -226,10 +229,46 @@ func checkArguments(args ArgNodesCoordinator) error {
...
+// it has to be used under mutex
+func (ihnc *indexHashedNodesCoordinator) getNodesConfig(epoch uint32) (*epochNodesConfig, bool) {
+	nodesConfig, ok := ihnc.nodesConfig[epoch]
+	if ok {
+		return nodesConfig, ok
+	}
+
+	value, ok := ihnc.nodesConfigCacher.Get([]byte(fmt.Sprint(epoch)))
```
Comment (AdoAdoAdo) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:247]: I think we can add in the cache the object directly unmarshalled so we don't need to unmarshal it if found.
if not found, we could look into the nodes coordinator registry storer (bootstrap data storage I think) with a GetFromEpoch, and if found unmarshal it and add it to the LRU cache.

then maybe rename also the method to getNodesConfigWithCacheUpdate

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 3]
```diff
@@ -281,6 +320,12 @@ func (ihnc *indexHashedNodesCoordinator) setNodesPerShards(
 	}
 
 	ihnc.nodesConfig[epoch] = nodesConfig
+	ncBytes, err := epochNodesConfigToRegistryBytes(epoch, nodesConfig)
+	if err != nil {
+		return err
+	}
+	ihnc.nodesConfigCacher.Put([]byte(fmt.Sprint(epoch)), ncBytes, len(ncBytes))
```
Comment (AdoAdoAdo) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:327]: no need to put into cacher if already put into nodesConfig[epoch] map

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 4]
```diff
@@ -226,10 +229,46 @@ func checkArguments(args ArgNodesCoordinator) error {
 	if args.NumStoredEpochs < minStoredEpochs {
 		return ErrInvalidNumberOfStoredEpochs
 	}
+	if check.IfNilReflect(args.NodesConfigCache) {
```
Comment (iulianpascalau) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:232]: Please add the definition of `IsInterfaceNil() bool` in the Cacher definition and use here `check.IfNil()`

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 2 above)
Comment (iulianpascalau) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:247]: Agree, Cachers can hold just about everything as value

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 4 above)
Comment (ssd04) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:232]: done

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 1 above)
Comment (ssd04) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:239]: done

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 2 above)
Comment (ssd04) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:247]: done

**Review by iulianpascalau — DISMISSED** [review]

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 5]
```diff
@@ -229,44 +229,53 @@ func checkArguments(args ArgNodesCoordinator) error {
...
+		if ok {
+			return enc, ok
+		}
 	}
 
-	registryBytes, ok := value.([]byte)
-	if !ok {
-		return nil, ok
+	ncInternalkey := append([]byte(common.NodesCoordinatorRegistryKeyPrefix), []byte(fmt.Sprint(epoch))...)
+	epochConfigBytes, err := ihnc.bootStorer.SearchFirst(ncInternalkey)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go]: changed to use getFromEpoch

File: sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go [hunk 6]
```diff
@@ -2564,4 +2564,55 @@ func TestIndexHashedNodesCoordinator_GetNodesConfig(t *testing.T) {
 		require.True(t, ok)
 		require.NotNil(t, nc)
 	})
+
+	t.Run("should work will old epochs", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go]: done

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**